src/openrct2/localisation/LocalisationService.cpp

```diff
@@ -0,0 +1,166 @@
+{
+    auto locale = std::string(LanguagesDescriptors[languageId].locale);
+    auto languageDirectory = _env->GetDirectoryPath(DIRBASE::OPENRCT2, DIRID::LANGUAGE);
+    auto languagePath = Path::Combine(languageDirectory, locale + ".txt");
+    return languagePath;
+}
+
+void LocalisationService::OpenLanguage(sint32 id, IObjectManager& objectManager)
+{
+    CloseLanguages();
+    if (id == LANGUAGE_UNDEFINED)
+    {
+        throw std::invalid_argument("id was undefined");
+    }
+
+    std::string filename;
+    if (id != LANGUAGE_ENGLISH_UK)
+    {
+        filename = GetLanguagePath(LANGUAGE_ENGLISH_UK);
+        _languageFallback = std::unique_ptr<ILanguagePack>(LanguagePackFactory::FromFile(LANGUAGE_ENGLISH_UK, filename.c_str()));
+    }
+
+    filename = GetLanguagePath(id);
+    _languageCurrent = std::unique_ptr<ILanguagePack>(LanguagePackFactory::FromFile(id, filename.c_str()));
+    if (_languageCurrent != nullptr)
+    {
+        _currentLanguage = id;
+        TryLoadFonts(*this);
+
+        // Objects and their localised strings need to be refreshed
+        objectManager.ResetObjects();
+    }
+    else
+    {
+        throw std::runtime_error("Unable to open language " + std::to_string(id));
+    }
+}
+
+void LocalisationService::CloseLanguages()
+{
+    _languageFallback = nullptr;
+    _languageCurrent = nullptr;
+    _currentLanguage = LANGUAGE_UNDEFINED;
+}
+
+std::tuple<rct_string_id, rct_string_id, rct_string_id> LocalisationService::GetLocalisedScenarioStrings(const std::string& scenarioFilename) const
+{
+    auto result0 = _languageCurrent->GetScenarioOverrideStringId(scenarioFilename.c_str(), 0);
+    auto result1 = _languageCurrent->GetScenarioOverrideStringId(scenarioFilename.c_str(), 1);
+    auto result2 = _languageCurrent->GetScenarioOverrideStringId(scenarioFilename.c_str(), 2);
+    return std::make_tuple(result0, result1, result2);
+}
+
+rct_string_id LocalisationService::GetObjectOverrideStringId(const char * identifier, uint8 index) const
+{
+    if (_languageCurrent == nullptr)
+    {
+        return STR_NONE;
+    }
+    return _languageCurrent->GetObjectOverrideStringId(identifier, index);
+}
+
+rct_string_id LocalisationService::AllocateObjectString(const std::string& target)
+{
+    auto stringId = _availableObjectStringIds.top();
+    _availableObjectStringIds.pop();
+    _languageCurrent->SetString(stringId, target);
+    return stringId;
+}
+
+void LocalisationService::FreeObjectString(rct_string_id stringId)
+{
+    if (stringId != STR_EMPTY)
+    {
+        if (_languageCurrent != nullptr)
+        {
+            _languageCurrent->RemoveString(stringId);
+        }
+        _availableObjectStringIds.push(stringId);
+    }
+}
+
+sint32 LocalisationService_GetCurrentLanguage()
+{
+    const auto& localisationService = GetContext()->GetLocalisationService();
+    return localisationService.GetCurrentLanguage();
+}
+
+bool LocalisationService_UseTrueTypeFont()
```

Comment:
These names look quite ugly? What the idea behind them? Are they temporary?

Reply:
Yes, they are deliberately ugly to make you not want to ever type them and change all existing ones asap.

Reply:
I'd put a `@todo` in, then.

Reply:
done